### ISSUE TITLE
Confirm ignored autoupdate exceptions work as expected

### ISSUE THREAD

**Issue by hepcat72** [Princeton-LSI-ResearchComputing/tracebase#554]
### Inspiration

TransactionManagementErrors are ignored in the MaintainedModel code.

### Description

There has been some concerns expressed over the catching of exceptions (related to unique constraint violations: IntegrityError and ForeignKeyViolation, and related to queries inside atomic blocks: transactionManagementError) which are turned into warnings.

On the one hand, the exception handling is justified in that it only affects fields that are being auto-updated and that any other fields will only throw exceptions of concern via the `super().save()` call, whose exceptions are *not* turned into warnings.  None of the warnings or exceptions currently are thrown in the test data.

On the other hand, the code is reasonably sophisticated that there could potentially exist a case that hasn't been accounted for and as such, there is concern that perhaps it doesn't behave as expected.

The recommended "feature" here is testing, discussion, and documentation.  Tests should be created to trigger these warnings.  A discussion should be had about the usage of queries to propagate child record updates in the event that auto-updates are triggered inside an atomic transaction.  And anything learned further about the cases which cause the errors should be better documented.

### Alternatives

Require that auto-updates not be done during loading and instead lazy-load the maintained field values.  See #553.

### Dependencies

none

### Comment

Note that the only reason queries are performed inside the auto-update code is so that related record propagations can traverse many-related models (1:M - from 1 record to many child records and M:M).  Traversals of M:1 and 1:1 can occur without hitting the database, so auto-updates that only traverse in that direction can be performed inside an atomic transaction.

#### Original PR comment:

        I'm still a bit uncomfortable about some of the existing catches/ignore/warns, but your proposal sounds like a reasonable plan forward. Unfortunately, it looks like this branch was rebased (or at least force-pushed somehow), so it's tough for me to see what changed since my last review. Rob, if you are confident that these changes are needed for things to work properly, then go ahead and consider this approved, assuming the issues are created for proper follow-up in the next milestone.

_Originally posted by @lparsons in https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/545#pullrequestreview-1136356506_

-----

### Assumptions

- List of assumptions that the code will not explicitly address/check
- E.g. We will assume input is correct (explaining why there is no validation)

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

### Limitations

- A list of things this work will specifically not do
- E.g. This feature will only handle the most frequent use case X

### Affected Components

A tentative list of anticipated repository items that will be changed, labeled
with "add", "delete", or "change".  One item per line.  (Mostly, this will be a
list of files.)

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

#### Interface Change description

Describe changes to usage. *E.g. GUI/command-line changes*

#### Code Change Description

Describe code changes planned for the feature. *(Pseudocode encouraged)*

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

**Comment by hepcat72** [comment]
I'm going to close this issue.  `TransactionManagementErrors` are no longer caught.  The catches were able to be removed after the update to Django 4.2, which added a new exception that shed light on what was going on with the `TransactionManagementErrors` and the misleading wording in its error about queries during an atomic transaction.  Clearly, queries during an atomic transaction are not a problem.  We have lots of queries that occur during atomic transactions without a problem.  In Django 4.2, a new case raised an exception that previously (in django 3.2) happily returned `None`, and that's when a newly created object's reverse relation was traversed (for example, in a property method), e.g. `animal_rec.samples...` - if `animal_rec` is being created in the atomic transaction, referencing the sample records that link to it cannot return anything.  In django 3.2, there was no exception.  Queries that traversed that reverse relation would just behave as if there was no result, i.e. as if it was a valid query.

That's just one example.  There were other cases in `MaintainedModel` where similar reverse relations were being traversed.  So, a refactor of `MaintainedModel` implemented a way to avoid traversing those queries when the model object does not exist in the database.  (A method was created, named something like `exists_in_db` (though I dan't recall the exact name).  That method was used to avoid making those queries in the `MaintainedModel` code.  **However**, `MaintainedModel`'s decorators could be used to decorate methods that perform auto-updates (during an atomic transaction).  In the past, before I better understood what was going on, I had concluded that the auto-updates just could not be inside the atomic transaction block, but now with the better understanding prompted by the new exception cases in Django 4.2, I was able to fix the decorated methods to perform the same check (i.e. see if the record exists in the DB before traversing the reverse relation), thus, with those fixes, I was able to remove the `TME` catches inside `MaintainedModel`.  In other words, if we write a new method to auto-update a field, and it improperly traverses a reverse relation without that object existing in the DB, it will raise an exception, as it should - meaning that the method's code should be fixed to take into account the possibility that the record does not yet exist in the DB.

Now, the thing to note, after making those fixes, is that mass auto-updates should not be run inside the same atomic transaction block where the records were created - because the autoupdates will result in incomplete updates (if its update relies on traversing a reverse relation).  SO - I added a separate atomic transaction block around the mass auto-update code.

## TL;DR

- `TransactionManagementError`s are no longer caught and ignored
- Methods that output auto-update values for fields must avoid traversing reverse relations when the record does not exist (if it does, a `TME` will be raised)
- Mass autoupdates CAN run inside an atomic block (and now do in their own block), but SHOULD run outside of the atomic transaction block where the records were created (and DO, when the `defer_autoupdates` decorator is applied to the loading code).

## Related issue to note

Note that there is still code that ignores `IntegrityError`s and `ValidationError`s (among other exceptions), however there are 2 things to note about that:

- Every exception ultimately is raised at the end of a load as an `AggregatedErrors` exception, causing a rollback.  (Previously, the `TME`s were ignored and a load without other errors were allowed to persist, but not anymore.)
- As a part of the current submission process refactor, any exception that occurs from a single record (or series of linked records) during a load, is rolled back.  This behavior thus far has been implemented for the tissues, compounds, treatments, sequences, tracers, and infusates loaders, and is planned in the loaders that remain to be refactored.
